src/org/rascalmpl/util/maven/SimpleRepositoryDownloader.java

```diff
@@ -50,46 +60,142 @@ public SimpleRepositoryDownloader(Repository repo, HttpClient client) {
+        if (expectedSha1 != null) {
+            return expectedSha1.equals(actualSha1);
+        }
+
+        String expectedMd5 = getChecksum(headers, Arrays.asList("x-checksum-md5", "x-goog-meta-checksum-md5"), url + ".md5");
+        if (expectedMd5 != null) {
+            return expectedMd5.equals(actualMd5);
+        }
+
+        // No checksum to check against
+        return true;
+    }
+
+    // Retrieve the checksum from one of the possible checksum headers. If not found, try to download it
+    private String getChecksum(HttpHeaders headers, List<String> checksumHeaders, String checksumUrl)
+        throws IOException, InterruptedException, URISyntaxException {
+        for (String headerName : checksumHeaders) {
+            Optional<String> value = headers.firstValue(headerName);
+            if (value.isPresent()) {
+                return value.get();
+            }
+        }
+
+        // Maybe we should return null on an IOException? That would mean no checksum checking in that case.
+        return downloadChecksum(createUri(repo.getUrl(), checksumUrl));
+    }
+
+    private String downloadChecksum(URI uri) throws IOException, InterruptedException {
+        var req = HttpRequest.newBuilder(uri).GET().build();
+        HttpResponse<String> result = client.send(req, BodyHandlers.ofString());
+        return result.statusCode() == 200 ? result.body() : null;
+    }
+
     private URI createUri(String url, String suffix) throws URISyntaxException {
         if (url.endsWith("/") && suffix.startsWith("/")) {
             suffix = suffix.substring(1);
         }
         return new URI(url + suffix);
     }
 
-    private void copyFileToTarget(Path target, boolean force, Path tempFile) throws IOException {
+    private boolean copyFileToTarget(Path target, boolean force, Path tempFile) throws IOException {
         var directory = target.getParent();
         if (Files.notExists(directory)) {
             Files.createDirectories(directory);
         }
         if (force || Files.notExists(target)) {
             Files.move(tempFile, target, StandardCopyOption.REPLACE_EXISTING);
+            return true;
+        }
+
+        return false;
+    }
+
+    private boolean writeToTarget(Path target, boolean force, String content) throws IOException {
+        var directory = target.getParent();
+        if (Files.notExists(directory)) {
+            Files.createDirectories(directory);
+        }
+        if (force || Files.notExists(target)) {
+            Files.writeString(target, content);
+            return true;
+        }
+
+        return false;
+    }
+
+    private void writeChecksumToTarget(Path path, String checksum) throws IOException {
+        if (checksum == null) {
+            Files.delete(path);
+        } else {
+            Files.write(path, checksum.getBytes(), StandardOpenOption.TRUNCATE_EXISTING, StandardOpenOption.CREATE);
```

Comment:
`getBytes` is depends on the system default to which encoding we get. So let's always add the specific encoding we intended. For example `StandardCharSets.UTF8`?